[JENKINS-55933] Add configuration to test slowdowns in the plugin

AbhyudayaSharma opened the pull request:
See [JENKINS-55933](https://issues.jenkins-ci.org/browse/JENKINS-55933).

AbhyudayaSharma commented:
Instead of adding all these files, we could also consider adding a git submodule. Please let me know.

oleg-nenashev commented:
@AbhyudayaSharma did you have a chance to create a JIRA ticket for it?

oleg-nenashev commented on src/test/performance/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+ENV JENKINS_UC_EXPERIMENTAL=https://updates.jenkins.io/experimental
+COPY plugins.txt /usr/share/jenkins/ref/plugins.txt
+COPY download-and-check-updates.sh /usr/local/bin/download-and-check-updates.sh
+RUN /usr/local/bin/download-and-check-updates.sh < /usr/share/jenkins/ref/plugins.txt
```
To keep the demo simple, I would remove this script and use `install-plugins.sh` from the parent image

AbhyudayaSharma replied:
Done!

oleg-nenashev commented on src/test/performance/agent/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+###
+# Custom agent build:
+# Image name: onenashev/demo-jenkins-maven-builder
```
Not needed, because the image will be taken from DockerHub anyway

oleg-nenashev commented on src/test/performance/download-and-check-updates.sh:
```diff
@@ -0,0 +1,304 @@
+#!/bin/bash -eu
```
The file can be removed entirely, see the `install-plugins.sh` comment

oleg-nenashev commented on src/test/performance/init_scripts/src/main/groovy/scripts/DevelopmentFolder.groovy:
```diff
@@ -0,0 +1,86 @@
+// Initializes the Development folder, which is fully configurable by the user
```
Is this folder used in your Demo?

oleg-nenashev commented on src/test/performance/init_scripts/src/main/groovy/scripts/DemoFolder.groovy:
```diff
@@ -0,0 +1,52 @@
+package scripts
```
Is this folder used in your Demo?

oleg-nenashev commented on src/test/performance/init_scripts/src/main/groovy/scripts/System.groovy:
```diff
@@ -0,0 +1,31 @@
+import hudson.security.csrf.DefaultCrumbIssuer
```
Is this folder used in your Demo?

oleg-nenashev commented on src/test/performance/jenkins.yaml:
```diff
@@ -0,0 +1,61 @@
+  authorizationStrategy:
+    roleBased:
+      roles:
+        global:
```
If you want to reproduce performance issues related to regular expressions & co, makes sense to add some project roles here

AbhyudayaSharma replied:
I've added the roles in ProductionFolder.groovy

oleg-nenashev commented on src/test/performance/pom.xml:
```diff
@@ -0,0 +1,40 @@
+  
+  <groupId>io.jenkins.systemgroovy.demo</groupId>
+  <artifactId>parent</artifactId>
+  <name>Jenkins Configuration-as-Code Demo</name>
```
Fix the name and the description?

AbhyudayaSharma replied:
Done

oleg-nenashev commented:
> Instead of adding all these files, we could also consider adding a git submodule. Please let me know.

@AbhyudayaSharma my recommendation would be to keep it as is, but to cleanup unused logic. It might be also possible to simplify the repository by extending the original image instead of copying it

AbhyudayaSharma commented on src/test/performance/README.md:
````diff
@@ -0,0 +1,60 @@
+docker run -d -v /var/run/docker.sock:/var/run/docker.sock -p 2376:2375 bobrik/socat TCP4-LISTEN:2375,fork,reuseaddr UNIX-CONNECT:/var/run/docker.sock
+```
+
+#### Developing Pipeline libraries
````
I've left the parts in the README for creating and running the image.

AbhyudayaSharma commented on src/test/performance/init_scripts/src/main/groovy/scripts/ProductionFolder.groovy:
```diff
@@ -0,0 +1,43 @@
+    def roleName = "myRole" + i
+    def userName = "user" + i
+    if (rand.nextBoolean()) {
+        strategy.doAddRole(RoleBasedAuthorizationStrategy.PROJECT, roleName,
```
I've added the roles here.

AbhyudayaSharma commented on src/test/performance/README.md:
````diff
@@ -0,0 +1,34 @@
+### Building image
+
+```shell
+docker build -t onenashev/demo-jenkins-config-as-code .
````
Should we be using a different image name?

oleg-nenashev replied:
yes, it would be great. Maybe you just want to update the makefile and to reference it from here

AbhyudayaSharma replied:
@oleg-nenashev I could push an image to my DockerHub account called something like `abhyudaya/jenkins-role-strategy-slowdown-config` and change the references here and in the Makefile. Should I continue with it?

I'm sorry for the delayed response. I had a few exams.

AbhyudayaSharma commented on src/test/performance/plugins.txt:
```diff
@@ -0,0 +1,27 @@
+matrix-auth:2.3
```
Should unnecessary plugins be removed?

oleg-nenashev replied:
Yes, it makes sense to do a cleanup. In this case the plugin is used as a dependency afaik

AbhyudayaSharma commented:
@oleg-nenashev I've made the changes and removed the unused files. I had a few questions and have left them above.

AbhyudayaSharma commented:
@oleg-nenashev I've removed the plugins to which I could find no code references to and have also removed the explicitly mentioned dependencies as they are automatically resolved when `install-plugins.sh` runs.

AbhyudayaSharma commented:
> @oleg-nenashev I could push an image to my DockerHub account called something like abhyudaya/jenkins-role-strategy-slowdown-config and change the references here and in the Makefile. Should I continue with it?

@oleg-nenashev A polite ping.

oleg-nenashev commented on src/test/performance/LICENSE.md:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2017 Oleg Nenashev
```
Needs to be updated or removed entirely. You also contributed to the code, so please add yourself as well

```suggestion
Copyright (c) 2017-2019 Oleg Nenashev, Abhyudaya Sharma and other contributors
```

AbhyudayaSharma commented:
@oleg-nenashev I've pushed the corresponding Docker image to DockerHub https://cloud.docker.com/u/abhyudaya/repository/docker/abhyudaya/jenkins-role-strategy-slowdown-config and have updated the Makefile and README accordingly.
